### PR TITLE
KEYCLOAK-15395. Removed unnecessary Account REST API endpoints

### PR DESCRIPTION
**Pull request by miquelsi**
KEYCLOAK-15395. Removed totp/remove (DELETE) and credentials/password (GET, POST) endpoints.

### CODE REVIEW

**Comment by miquelsi** [comment]
Please @vmuzikar, @ASzc, @abstractj, @slaskawi review at your earliest convenience.

**Comment by abstractj** [comment]
@ssilvert assuming that this may have an impact in the new account console, would you like to assign someone from your team to review?

**Comment by ssilvert** [comment]
> @ssilvert assuming that this may have an impact in the new account console, would you like to assign someone from your team to review?

I took a quick look.  As long as the UI tests pass it's OK with me.  Kicked off UI test build number 978 for this.

**Comment by thomasdarimont** [comment]
Removing those endpoints might break existing applications. This should at least be mentioned in the migration guide / changelog.
In general, I think it would be good to have a deprecation process in place where changes like this are announced first (per mailing list / blog, code annotated with @Deprecated) before the actual API removal is performed.

**Comment by vmuzikar** [comment]
@thomasdarimont This is still a preview (unsupported) API. To make it final and supported, we need to do some clean up. ;) I believe it is ok to do this change. @stianst?

**Comment by abstractj** [comment]
> @thomasdarimont This is still a preview (unsupported) API. To make it final and supported, we need to do some clean up. ;) I believe it is ok to do this change. @stianst?

@thomasdarimont thanks for your feedback but I'm kinda of agree with @vmuzikar here. The account REST API is still marked as preview, which means that things may change and will change, until we flip the switch.

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (abstractj) [inline services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java:343]: We have some unused imports which are safe to remove like:

```
import org.keycloak.credential.PasswordCredentialProvider;
import org.keycloak.credential.PasswordCredentialProviderFactory;
import org.keycloak.events.EventType;
import javax.ws.rs.POST;
```

We can also remove:
```
private final EventBuilder event;
```

**Comment by ssilvert** [comment]
> > @ssilvert assuming that this may have an impact in the new account console, would you like to assign someone from your team to review?
> 
> I took a quick look. As long as the UI tests pass it's OK with me. Kicked off UI test build number 979 for this.

The tests on the new account console look fine.

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (miquelsi) [reply services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java:337]: Good catch. Thanks, I removed that.

**Review by abstractj — DISMISSED** [review]
@miquelsi LGTM, will wait until @vmuzikar review so we can merge.

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java
Comment (vmuzikar) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:404]: Just a very small nit pick.
```java
import org.keycloak.credential.CredentialModel;
```
is also unused now.

**Review by vmuzikar — COMMENTED** [review]
Thank you @miquelsi! Changes look good, found one very small nit pick – I think it's safe to merge as is.

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java
Comment (miquelsi) [reply services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:404]: Easy to fix, so I removed it :-)

**Review by vmuzikar — APPROVED** [review]

**Review by abstractj — APPROVED** [review]

**Comment by abstractj** [comment]
Merged, thanks @miquelsi

**Comment by MatCuk** [comment]
Hi,

can any give me more infromation what to use now to change password over REST call?
I'm also asking on discord with not answers : 
https://keycloak.discourse.group/t/keycloak-account-api-password-endpoint-removed-why/5274

So my use case is I have my FE app which would call my BE and my BE would forward change password request to keycloak.

**Comment by loudmouth** [comment]
Is there a plan on replacing the functionity to allow users to updat their own passwords as was previously enabled with a  `POST`  to the `account/credentials/password` endpoint?

I know it was in preview, so it is definitely fair game that it was removed, but it _was_ nice to provide our account management experience in our react app with custom UI and rely upon that endpoint. I fully expected as a preview API that it might change, but did not expect that it would simply be removed.

**Comment by vmuzikar** [comment]
@loudmouth @MatCuk AFAIK there're no plans to support credentials modifications via REST API as it was considered a security risk. An app that is secured by Keycloak should not have access to the user credentials.

As an alternative, you can use [Application Initiated Actions (AIA)](https://github.com/keycloak/keycloak-community/blob/master/design/application-initiated-actions.md) to trigger a browser flow where user will be prompted to update their password. This is [what the new Account Console uses](https://github.com/keycloak/keycloak/blob/e56bd9d8b8cbaf4600e6bcc5e28c962eb561d74c/themes/src/main/resources/theme/keycloak.v2/account/src/app/util/AIACommand.ts). Another option is to [use Admin REST API](https://www.keycloak.org/docs-api/12.0/rest-api/index.html#_resetpassword).

**Comment by loudmouth** [comment]
alrighty thank you @vmuzikar 

I realize that i may have stumbled upon what is supposed to be a private pull request so I appreciate you taking the time to answer.

It's very nice of you to point me to some documentation as I had been struggling to find the relevant information.

**Comment by mapkyca** [comment]
Hmm... this is critical functionality for us. Are there plans to build an equivalent, or is the concept just being dropped?

**Comment by vmuzikar** [comment]
@mapkyca Hi, there are no plans for adding/resetting credentials via REST APi.

**Comment by abascivan** [comment]
@vmuzikar Hi. I tried the example of using AIA that is described on the link you posted but I can't make it work.

If I understand well for example if we want to user to update his profile we should redirect him to an URL that looks like:

http://localhost:8050/auth/realms/myreal/protocol/openid-connect/auth?response_type=code&client_id=myapp&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2F&kc_action=update_profile

But When I do so it gives me error. Is the URL that I use seems correct ?

**Comment by sthenault** [comment]
@vmuzikar @abascivan I can't make it work either. Using similar url: 

- the fist time I get the same result as @abascivan, it prompt me the login page then go back to my application (even though I was already logged)
- the second time, I get an error. According to logs, I get a null pointer exception at https://github.com/keycloak/keycloak/blob/12.0.4/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocol.java#L386 - it seems that the requiredActionProvider is null while this is not expected.

Any clue about this? bug, misconfiguration, bad url?
BTW is there some documentation about AIA beside the md file you mentioned above?

thx!

**Comment by vmuzikar** [comment]
@abascivan @sthenault I'm afraid it's not that easy to generate the URL "manually". Should be done through the adapter but AFAIK only the JS adapter supports in at the moment. IMHO a contribution for adding AIA support to other adapters would be most welcome! :) You could discuss it on the [keycloak-dev list](https://groups.google.com/g/keycloak-dev).

Please, if you have any follow-up usage questions, address them to the [keycloak-user list](https://groups.google.com/g/keycloak-user) for better visibility to other users and team members. Thank you!

**Comment by barashnikov** [comment]
@vmuzikar Hello, I'm trying to do that via JS adapter but I have no clue to how to do that ! 

By any chance did you get an example of an implementation for change changing password ?

Thank you !

**Comment by vmuzikar** [comment]
@barashnikov I think our Account Console could [give you a few hints](https://github.com/keycloak/keycloak/blob/master/themes/src/main/resources/theme/keycloak.v2/account/src/app/util/AIACommand.ts). ;)

**Comment by barashnikov** [comment]
@vmuzikar Thank you for your fast response.

I would like to show you my code : 

async componentDidMount() {

    if (keycloakEnabled) {
      var keycloak = Keycloak('/keycloak.json');
      await keycloak.init({ onLoad: 'login-required' }).then(authenticated => {
        sessionStorage.setItem('token', keycloak.token);

        keycloak.loadUserInfo().then(function (data) {
          sessionStorage.setItem('Locale', data.locale)
          sessionStorage.setItem('username', data.preferred_username)
          sessionStorage.setItem('name', data.name)
        });


        this.setState({ keycloak: keycloak, authenticated: authenticated, props: keycloak });
        setInterval(() => {
          if (keycloak.isTokenExpired) {
            keycloak.updateToken(-1).then((refresh) => {

              sessionStorage.setItem('token', keycloak.token);
            })

          }
        }, 1000 * 60 * 5);

      });
    }
  }

As you can see, I'm currently using some methods of keycloak for retrieving profile info for an user.

But I couldn't find anything about change password or maybe I didn't search well.

Thx again !

**Comment by vmuzikar** [comment]
@barashnikov Please see the link I posted in my previous comment. You basically need to call `kecloak.login()` with `UPDATE_PASSWORD` action.

If you have further questions, please use [keycloak-user list](https://groups.google.com/g/keycloak-user). An old PR is not the best place to discuss such things. Thank you.

**Comment by anwar-bs** [comment]
This is a sample url that prompts for password update UI, tested in 15.0.2

https://your.keycloak.url.com/auth/realms/YOUR_REALM_NAME/protocol/openid-connect/auth?client_id=account-console&redirect_uri=https%3A%2F%2Fgoogle.com&response_type=code&scope=openid&kc_action=UPDATE_PASSWORD&code_challenge=v3edtsDSqpv8xCVtvXU1KxuJZ1soLQp3ZzOz2SyhjMo&code_challenge_method=plain

code_challenge_method which is set in keycloak admin panel for **Account-console => settings => Advanced settings => Proof Key for Code Exchange Code Challenge Method** must be used (plain or S256)

**Comment by saghayar** [comment]
> This is a sample url that prompts for password update UI, tested in 15.0.2
> 
> https://your.keycloak.url.com/auth/realms/YOUR_REALM_NAME/protocol/openid-connect/auth?client_id=account-console&redirect_uri=https%3A%2F%2Fgoogle.com&response_type=code&scope=openid&kc_action=UPDATE_PASSWORD&code_challenge=v3edtsDSqpv8xCVtvXU1KxuJZ1soLQp3ZzOz2SyhjMo&code_challenge_method=plain
> 
> code_challenge_method which is set in keycloak admin panel for **Account-console => settings => Advanced settings => Proof Key for Code Exchange Code Challenge Method** must be used (plain or S256)

Thank you @anwar-bs , the link you posted worked fine .

**Comment by saghayar** [comment]
> This is a sample url that prompts for password update UI, tested in 15.0.2
> 
> https://your.keycloak.url.com/auth/realms/YOUR_REALM_NAME/protocol/openid-connect/auth?client_id=account-console&redirect_uri=https%3A%2F%2Fgoogle.com&response_type=code&scope=openid&kc_action=UPDATE_PASSWORD&code_challenge=v3edtsDSqpv8xCVtvXU1KxuJZ1soLQp3ZzOz2SyhjMo&code_challenge_method=plain
> 
> code_challenge_method which is set in keycloak admin panel for **Account-console => settings => Advanced settings => Proof Key for Code Exchange Code Challenge Method** must be used (plain or S256)

Thank you @anwar-bs , the link you posted worked fine .

**Comment by IvanDugandzic** [comment]
> @loudmouth @MatCuk AFAIK there're no plans to support credentials modifications via REST API as it was considered a security risk. An app that is secured by Keycloak should not have access to the user credentials.
> 
> As an alternative, you can use [Application Initiated Actions (AIA)](https://github.com/keycloak/keycloak-community/blob/master/design/application-initiated-actions.md) to trigger a browser flow where user will be prompted to update their password. This is [what the new Account Console uses](https://github.com/keycloak/keycloak/blob/e56bd9d8b8cbaf4600e6bcc5e28c962eb561d74c/themes/src/main/resources/theme/keycloak.v2/account/src/app/util/AIACommand.ts). Another option is to [use Admin REST API](https://www.keycloak.org/docs-api/12.0/rest-api/index.html#_resetpassword).

As i can see in latest [Admin API Docs](https://www.keycloak.org/docs-api/22.0.3/rest-api/index.html)  
action `PUT /admin/realms/{realm}/users/{id}/reset-password` is still available. Can we count on this API in future versions since regarding security it can also be considered as "credentials modifications via REST API" just like actions removed in this PR?

**Pull request merged**
